lxd/instance.go

```diff
@@ -174,6 +174,137 @@ func instanceCreateFromImage(s *state.State, r *http.Request, img *api.Image, ar
 	return inst, nil
 }
 
+func instanceRebuildFromImage(s *state.State, r *http.Request, inst instance.Instance, img *api.Image, op *operations.Operation) error {
+	// Validate the type of the image matches the type of the instance.
+	imgType, err := instancetype.New(img.Type)
+	if err != nil {
+		return err
+	}
+
+	if imgType != inst.Type() {
+		return fmt.Errorf("Requested image's type %q doesn't match instance type %q", imgType, inst.Type())
+	}
+
+	// Check if the image is available locally or it's on another member.
+	// Ensure we are the only ones operating on this image. Otherwise another instance created at the same
+	// time may also arrive at the conclusion that the image doesn't exist on this cluster member and then
+	// think it needs to download the image and store the record in the database as well, which will lead to
+	// duplicate record errors.
+	unlock := imageOperationLock(img.Fingerprint)
+
+	nodeAddress, err := s.DB.Cluster.LocateImage(img.Fingerprint)
+	if err != nil {
+		unlock()
+		return fmt.Errorf("Locate image %q in the cluster: %w", img.Fingerprint, err)
+	}
+
+	if nodeAddress != "" {
+		// The image is available from another node, let's try to import it.
+		err = instanceImageTransfer(s, r, inst.Project().Name, img.Fingerprint, nodeAddress)
+		if err != nil {
+			unlock()
+			return fmt.Errorf("Failed transferring image %q from %q: %w", img.Fingerprint, nodeAddress, err)
+		}
+
+		// As the image record already exists in the project, just add the node ID to the image.
+		err = s.DB.Cluster.AddImageToLocalNode(inst.Project().Name, img.Fingerprint)
+		if err != nil {
+			unlock()
+			return fmt.Errorf("Failed adding transferred image %q record to local cluster member: %w", img.Fingerprint, err)
+		}
+	}
+
+	unlock() // Image is available locally.
+
+	instConfig := inst.LocalConfig()
+
+	// Reset the "image.*" keys.
+	for k := range instConfig {
+		if strings.HasPrefix(k, "image.") {
+			delete(instConfig, k)
+		}
+	}
+
+	for k, v := range img.Properties {
+		instConfig[fmt.Sprintf("image.%s", k)] = v
+	}
+
+	// Reset relevant volatile keys.
+	instConfig["volatile.base_image"] = img.Fingerprint
+	delete(instConfig, "volatile.idmap.next")
+	delete(instConfig, "volatile.last_state.idmap")
+	instConfig["volatile.uuid.generation"] = instConfig["volatile.uuid"]
+
+	err = inst.Rebuild(img.Fingerprint, op)
+	if err != nil {
+		return fmt.Errorf("Failed rebuilding instance from image: %w", err)
+	}
+
+	err = s.DB.Cluster.Transaction(context.TODO(), func(ctx context.Context, tx *db.ClusterTx) error {
+		object, err := dbCluster.GetInstance(ctx, tx.Tx(), inst.Project().Name, inst.Name())
```

Comment:
By pushing down this logic into `Rebuild` we will avoid needing to perform this extra query as the ID will already be loaded and available inside the instance.